src/upgrade.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""In-place upgrades
+
+Based off specification: DA058 - In-Place Upgrades - Kubernetes v2
+(https://docs.google.com/document/d/1tLjknwHudjcHs42nzPVBNkHs98XxAOT2BXGGpP7NyEU/)
+"""
+
+import abc
+import copy
+import enum
+import json
+import logging
+import pathlib
+import typing
+
+import ops
+import poetry.core.constraints.version as poetry_version
+from charms.opensearch.v0.opensearch_distro import OpenSearchDistribution
+
+logger = logging.getLogger(__name__)
+
+PEER_RELATION_ENDPOINT_NAME = "upgrade-version-a"
```

Comment:
Could we rename this to include `upgrade` in it?

Reply:
It's fine, but keep in mind, another developer may later during a refactor remove the module name on the reference - and then this might be prone to confusions and strange errors if reused elsewhere thinking it was `PeerRelationName`